### PR TITLE
back/jun3/testing/ Add testing questions and resources

### PR DESCRIPTION
**Pull request by stanislav-az**

### CODE REVIEW

**Comment by Znack** [comment]
А у нас же есть на фронтенд вопросы по тестированию :) тут по большей части все это совпадает. Может как-нибудь тут дать ссылку просто на фронтендерскую часть?

File: backend/junior-3/testing.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  * What are the advantages of TDD?
+  * What are the limitations of TDD?
+* What are test doubles: dummies, fakes, stubs, spies, mocks?
+* What are test fixtures?
+* What are edge cases? Why do we test them?
+* What time do you think you should spend on a one programming cycle: `writing tests -> coding changes -> refactoring`?
+* Why does the speed of running all tests matter?
+* How to make the reason of test failure obvious?
+* What are integration tests?  
+How do unit-tests differ from integration tests?
```
Comment (olgaklimenko) [inline backend/junior-3/testing.md]: This string lacks bullet list marker. And in some strings below too

**Comment by stanislav-az** [comment]
> А у нас же есть на фронтенд вопросы по тестированию :) тут по большей части все это совпадает. Может как-нибудь тут дать ссылку просто на фронтендерскую часть?

Придется тогда на русском эту часть сделать :) Если не принципиально, можно тогда тут просто оставить ссылки на jun-1, jun-2, jun-3 оттуда.

**Comment by Znack** [comment]
Да, просто ссылки оставить, на фронте как нибудь переведем :)

File: backend/junior-3/testing.md (hunk 1 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: The questions would be under the same bullet marker https://github.com/fullstack-development/developers-roadmap/pull/130#discussion_r357901269

**Comment by stanislav-az** [comment]
Ссылки сюда поставил к фронту)

File: backend/junior-3/testing.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+* What time do you think you should spend on a one programming cycle: `writing tests -> coding changes -> refactoring`?
+* Why does the speed of running all tests matter?
+* How to make the reason of test failure obvious?
+* What are integration tests?
+  * How do unit-tests differ from integration tests?
+  * Why it is more difficult to write and support integration tests?
+* What is test covering?
+  * How to find a balance between a number of tests and test covering?
+  What problems does the pursuit of 100% covering bring?
+* How do tests help in application construction?
```
Comment (Znack) [inline backend/junior-3/testing.md]: How do tests help while designing software architecture?

File: backend/junior-3/testing.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Testing for Junior-3
+
+This section is a translation of the following questions lists:
+
+* [front/jun-1](/frontend/junior-1/testing.md)
+* [front/jun-2](/frontend/junior-2/testing.md)
+* [front/jun-3](/frontend/junior-3/testing.md)
+
+## Questions
```
Comment (Znack) [inline backend/junior-3/testing.md]: нет вопросов:
* про сайд эффекты "Почему сайд-эффекты ухудшают тестируемость кода?", 
* Стоит ли тестировать приватные методы класса? Почему?
* Нужно ли автоматизированно тестировать сами тесты?
* Что такое heisenbug? Какие есть способы их дебаггинга?
* В чём преимущества правила одной концепции на один unit-тест?
* В чём преимущества изолированных друг от друга тестов?

Comment (stanislav-az) [reply backend/junior-3/testing.md]: Ну это был выборочный перевод, про приватные методы класса интересный вопрос, но не про хаскель. Можно аналогию, конечно, провести. Кстати, тут какая? С тем, что можно внутренние функции в модуле не экспортировать?

Comment (stanislav-az) [reply backend/junior-3/testing.md]: Добавлю их все, зря не стал переводить

Comment (Znack) [reply backend/junior-3/testing.md]: Ага, если не получается тестировать внутренние функции модуля, то как их тестировать?

Comment (stanislav-az) [reply backend/junior-3/testing.md]: Они тестируются через внешние функции которые их используют внутри

Comment (Znack) [reply backend/junior-3/testing.md]: Ну вот такой вопрос с таким предполагаемым ответом тут и должен быть :)

File: backend/junior-3/testing.md [hunk 4]
```diff
@@ -11,8 +11,11 @@ This section is a translation of the following questions lists:
 * What is automated testing, and how does it differ from manual?
   * What is the regression bug problem and how automated testing helps solving it?
   * What stages of manual testing are omitted by using automated testing instead?
+  * Is automated testing automatic tests an essential requirement?
```
Comment (Znack) [inline backend/junior-3/testing.md]: здесь неочевидная формулировка, можешь предложить что-нибудь корректней?

**Comment by Znack** [comment]
@performanceArtist @chmnkh можете плиз тоже глянуть мр с переводом вопросов про тестируемость?

**Comment by chmnkh** [comment]
да

File: backend/junior-3/testing.md (hunk 4 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: * Do you need to automatically test the automatic tests?
Надеюсь говорят   "Do you need"

File: backend/junior-3/testing.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Testing for Junior-3
+
+This section is a translation of the following questions lists:
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: Я предлагаю такие штуки как-то оформлять отдельно, типа там курсивом или цитатой (предлагаю потыкаться и решить), а дальше вопросы как обычно писать безо всяких Questions

File: backend/junior-3/testing.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+* [front/jun-1](/frontend/junior-1/testing.md)
+* [front/jun-2](/frontend/junior-2/testing.md)
+* [front/jun-3](/frontend/junior-3/testing.md)
+
+## Questions
+
+* What is automated testing, and how does it differ from manual?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: я бы после manual добавил существительное, не уверен, что правильно без него

File: backend/junior-3/testing.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+Are these tests required?
+* What is a heisenbug?
+What are the methods of debugging it?
+
+#### Resources
+
+* [A quick guide to Manual Testing Vs Automated Testing](https://reqtest.com/testing-blog/manual-testing-vs-automated-testing/)
+* [Зачем нужны юнит-тесты](https://tproger.ru/translations/unit-tests-purposes/)
+* [Как, используя TDD, сократить время разработки](https://www.simbirsoft.com/blog/razrabotka-cherez-testirovanie-polza-i-vred/)
+* [Об использовании модульных тестов и TDD](https://eax.me/unit-testing/)
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: все английское, а ресурсы русские, это норм?

File: backend/junior-3/testing.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+This section is a translation of the following questions lists:
+
+* [front/jun-1](/frontend/junior-1/testing.md)
+* [front/jun-2](/frontend/junior-2/testing.md)
+* [front/jun-3](/frontend/junior-3/testing.md)
+
+## Questions
+
+* What is automated testing, and how does it differ from manual?
+  * What is the regression bug problem and how automated testing helps solving it?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: 1. что значит solve a bug?
2. почему тут перед and нет запятой, а выше - есть?

how does automated testing help to deal with it?

File: backend/junior-3/testing.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  * What stages of manual testing are omitted by using automated testing instead?
+  * Do you need to automatically test the automatic tests?
+* What is a unit-test?
+  * Why is it useful to write unit-tests in groups?
+  * What are the pros and cons of testing one smallest conception for a one unit-test?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: the smallest по определению одно, не понятно что за one

File: backend/junior-3/testing.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  * What are the pros and cons of making tests isolated and self-contained?
+* How do tests help novice programmers?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: наверное не новичкам-программмистам, а новым людям на проекте?

File: backend/junior-3/testing.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+* How do tests help in writing project documentation?
+* How do tests help with receiving feedback from the coding process?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: такое ощущуние что процесс какой-то там фидбэк даст, не?

File: backend/junior-3/testing.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+Why does the feedback matter?
+* How do tests help refactoring process?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: help in? during? ну т.е. сйечас так, как будто они процессу помогают, а по идее наверное должны помогать во время процесса людям (их явно не упоминать конечно можно в вопросе), как-то так

File: backend/junior-3/testing.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+* What is TDD?
+  * Why this approach to coding exists?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: меня смущает to coding, я оно вообще избыточно, я бы убрал эти слова

File: backend/junior-3/testing.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  * What are the advantages of TDD?
+  * What are the limitations of TDD?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: мне кажется вопрос про минусы должен отдельно стоять, а не только вопрос про ограничения

странно спрашивать про плюсы, но про минусы явно не спрашиватт

File: backend/junior-3/testing.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+* What are test doubles: dummies, fakes, stubs, spies, mocks?
+* What are test fixtures?
+* What are edge cases? Why do we test them?
+* What time do you think you should spend on a one programming cycle: `writing tests -> coding changes -> refactoring`?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: тут таки артикль не нужен перед one?

File: backend/junior-3/testing.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+* Why does the speed of running all tests matter?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: зачем тут the?

File: backend/junior-3/testing.md [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+* How to make the reason of test failure obvious?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: почему the a не a?

Comment (chmnkh) [reply backend/junior-3/testing.md]: перед failure почему ничего?

File: backend/junior-3/testing.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+* What are integration tests?
+  * How do unit-tests differ from integration tests?
+  * Why it is more difficult to write and support integration tests?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: is it

Comment (chmnkh) [reply backend/junior-3/testing.md]: сложнее относительно чего? юнит тестов? тогда надо после верхнего вопроса сразу писать, либо писать тут юнит тесты, потому что явно контекст не задан

File: backend/junior-3/testing.md [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+* What is test covering?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: coverage же всегда было вроде?

File: backend/junior-3/testing.md [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+  * How to find a balance between a number of tests and test covering?
+  What problems does the pursuit of 100% covering bring?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: я сомневаюсь, что есть такое выражение как bring a broblem в том смысле, котором тут оно должно быть

File: backend/junior-3/testing.md (hunk 7 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Раньше никто не жаловался) Я думаю, что норм, так как ресурсы это всего лишь рекомендации к прочтению.

File: backend/junior-3/testing.md (hunk 8 above)
Comment (kelizarov) [reply backend/junior-3/testing.md]: По-идеи не должно быть запятой)

File: backend/junior-3/testing.md (hunk 5 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Ну неплохая идея, я попробовал изобрести тут какой-то стиль

File: backend/junior-3/testing.md (hunk 6 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Правильно, но пусть будет дополнительная точность)

File: backend/junior-3/testing.md (hunk 8 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Правильно helps solving, или helps to solve.

Comment (stanislav-az) [reply backend/junior-3/testing.md]: Насчет запятой, вообще какой-то сложный неоднозначный случай. Тут просто слово it ссылается на разные вещи, как-то интуитивно не хочется туда ставить запятую)

File: backend/junior-3/testing.md (hunk 10 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Ага, так лучше

Comment (stanislav-az) [reply backend/junior-3/testing.md]: Новички уже туда включены

File: backend/junior-3/testing.md (hunk 8 above)
Comment (kelizarov) [reply backend/junior-3/testing.md]: Тут скорее всего запятая не нужна, так как `it` ссылается на контекст превой части предложения, а именно `automated testing`. Если бы вторая часть содержала подлежащее и сказуемое и не была связана, то запятая нужна была. Я по фасту чекнул правила тут https://www.businessinsider.com/a-guide-to-proper-comma-use-2013-9 поэтому не могу быть на 100% увереным быть.

File: backend/junior-3/testing.md (hunk 11 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Ну да, процесс дает фидбек, но более точно - тесты как часть этого процесса

File: backend/junior-3/testing.md (hunk 12 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Ага, during тогда

File: backend/junior-3/testing.md (hunk 13 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: А, ну to development бы больше зашло, но лучше убрать

File: backend/junior-3/testing.md (hunk 15 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Нужен, так как тут говорится в общем о каком-то цикле из циклов вида `writing tests -> coding changes -> refactoring`

File: backend/junior-3/testing.md (hunk 16 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Вроде не нужен

File: backend/junior-3/testing.md (hunk 17 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: _the_ reason of _a_ test failure тогда

File: backend/junior-3/testing.md (hunk 19 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: image

File: backend/junior-3/testing.md (hunk 20 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Так-то есть https://context.reverso.net/%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4/%D0%B0%D0%BD%D0%B3%D0%BB%D0%B8%D0%B9%D1%81%D0%BA%D0%B8%D0%B9-%D1%80%D1%83%D1%81%D1%81%D0%BA%D0%B8%D0%B9/what+brings#%D1%87%D1%82%D0%BE+%D0%BF%D1%80%D0%B8%D0%BD%D0%BE%D1%81%D0%B8%D1%82+%D0%BD%D0%B0%D0%BC последний пример тут

File: backend/junior-3/testing.md (hunk 8 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md]: Круто) Только ссылается наверное на `regression bug`
> Тут скорее всего запятая не нужна, так как `it` ссылается на контекст превой части предложения, а именно `automated testing`. Если бы вторая часть содержала подлежащее и сказуемое и не была связана, то запятая нужна была. Я по фасту чекнул правила тут https://www.businessinsider.com/a-guide-to-proper-comma-use-2013-9 поэтому не могу быть на 100% увереным быть.

**Comment by Znack** [comment]
@olgaklimenko @chmnkh можете глянуть плиз на след неделе?

File: backend/junior-3/testing.md (hunk 11 above)
Comment (chmnkh) [reply backend/junior-3/testing.md]: ну не, "тесты дают фидбэк" и "процесс кодинга дает фидбэк" это не одно и то же, одно не точнее другого, а разные по смыслу фразы, согласись)

Comment (chmnkh) [reply backend/junior-3/testing.md]: вообще я бы в формулировке уточнил что за фидбэк, потому что как-то сильно по-общему выглядит

File: backend/junior-3/testing.md (hunk 14 above)
Comment (chmnkh) [reply backend/junior-3/testing.md]: а limitations не входят в cons выше?)

File: backend/junior-3/testing.md (hunk 20 above)
Comment (chmnkh) [reply backend/junior-3/testing.md]: image

с Happiness что ли? найдешь с problem?) я не могу)

чтобы наверняка, я бы написал `cause`

и вне зависимости от глагола, который мы выберем, я бы еще написал пере ним `can`, чтобы корректнее была формулировка, типа `can cause`

File: backend/junior-3/testing.md [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+* Why does speed of running all tests matter?
+* How to make the reason of a test failure obvious?
+* What are integration tests?
+  * How do unit-tests differ from integration tests?
+  * Why is it more difficult to write and support integration tests compared with unit tests?
+* What is test coverage?
+  * How to find a balance between a number of tests and test coverage?
+  What problems does the pursuit of 100% covering bring?
+* How do tests help while designing software architecture?
+Why testability of a module is a good indicator of clear code?
```
Comment (chmnkh) [inline backend/junior-3/testing.md]: как будто чаще в таких случаях `clean` используют? я не уверен, но если бы я увидел `clear` я бы подумал в первую очередь что код понятный, а `clean` - уже чистый, качественный и всякое такое)

File: backend/junior-3/testing.md (hunk 11 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md:19]: соглашаюсь)

Comment (stanislav-az) [reply backend/junior-3/testing.md:19]: но вопрос непонятный какой-то, в оригинале звучит так: 
>* Как тесты ускоряют получение фидбека от проделанной работы и почему этот фидбек важен?

Comment (stanislav-az) [reply backend/junior-3/testing.md:19]: то есть тут тесты дают фидбек от кодинга

File: backend/junior-3/testing.md (hunk 14 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md:25]: я думаю что входят

File: backend/junior-3/testing.md (hunk 20 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md:37]: окей, разумно

File: backend/junior-3/testing.md (hunk 11 above)
Comment (stanislav-az) [reply backend/junior-3/testing.md:19]: хотя ладно, я понял вопрос, и перевел более точно

**Comment by stanislav-az** [comment]
@Znack @chmnkh совсем забыл что этот пулл реквест еще не замержили :) 
Исправил тут все.
Можно замержить и потом выносить в общее как @olgaklimenko уже сделала с гитом.

**Comment by Znack** [comment]
Стас, тут кстати за это время появились переводы этих вопросов от переводчиков-фрилансеров, которые нам топики переводили для англоязычных читателей карты, можешь заценить?

https://github.com/fullstack-development/developers-roadmap/pull/196/files

https://github.com/fullstack-development/developers-roadmap/pull/209/files

— вот тут файлы testing.md можно смотреть :) Мне кажется, там есть некоторые варианты, которые чутка лучше выражают мысль (нпример про ввод новичков в проект использование слова onboarding (хотя может это термин и может только запутать тех кто английского не знает) или про то какие этапы ручного тестирования можно избежать). Что думаешь, можем какие-то формулировки оттуда украсть? И можешь ли там в PR предложить какие-нибудь улучшения? :)

**Comment by stanislav-az** [comment]
> появились переводы этих вопросов от переводчиков-фрилансеров

А зачем тогда этот пулл реквест? Надо было его закрыть, а то я тут исправляю, а это уже не актуально...

**Comment by Znack** [comment]
Сложный вопрос :) По идее мы можем взять и текущую версию вопросов на фронтенде и просто напрямую переводы в Shared вынести. А можно и твои вопросы просто взять, а те переводы отменить :)

Но я больше сейчас склоняюсь, что нам всё-таки лучше деление тестирования на несколько уровней оставить, чтобы слишком большим третьего джуна сделать :)

**Comment by stanislav-az** [comment]
Я тоже за то чтобы оставить разделение на уровни, и за то чтобы вынести в Shared. Не помню с какой стати мы решили вообще все в один уровень слепить.

**Comment by Znack** [comment]
Окей, ну давай пока тут закроем. А можешь плиз тогда глянуть те два PR-а и оставить свои комментарии на эти темы хотя бы?

**Comment by stanislav-az** [comment]
Я посмотрел PRы, там перевод более близкий к оригиналу, у меня был вольный :)

**Comment by Znack** [comment]
Ну тебе более близкий к оригиналу нравится или вольный? :)

**Comment by stanislav-az** [comment]
Я думаю что принципиальной разницы нет
